applications/web-browser/controller.h

```diff
@@ -0,0 +1,83 @@
+#ifndef CONTROLLER_H
+#define CONTROLLER_H
+
+#include <QObject>
+#include <QImage>
+#include <QQuickItem>
+#include <QSettings>
+#include <QTimer>
+
+#include <epframebuffer.h>
+#include <liboxide.h>
+
+#include <QKeyEvent>
+#include <QMouseEvent>
+
+#define MOLD_SETTINGS_VERSION 1
+
+enum State { Normal, PowerSaving };
+enum BatteryState { BatteryUnknown, BatteryCharging, BatteryDischarging, BatteryNotPresent };
+enum ChargerState { ChargerUnknown, ChargerConnected, ChargerNotConnected, ChargerNotPresent };
+enum WifiState { WifiUnknown, WifiOff, WifiDisconnected, WifiOffline, WifiOnline};
+
+class Controller : public QObject {
+    Q_OBJECT
+public:
+    Controller(QObject* parent)
+    : QObject(parent), settings(this), applications() {
+        settings.sync();
+        auto version = settings.value("version", 0).toInt();
+        if(version < MOLD_SETTINGS_VERSION){
+            migrate(&settings, version);
+        }
+    }
+    ~Controller(){}
+
+    Q_INVOKABLE void startup(){
+        qDebug() << "Running controller startup";
+        QTimer::singleShot(10, [this]{
+            setState("loaded");
+        });
+    }
+    QString state() {
+        if(!getStateControllerUI()){
+            return "loading";
+        }
+        return stateControllerUI->property("state").toString();
+    }
+    void setState(QString state){
+        if(!getStateControllerUI()){
+            throw "Unable to find state controller";
+        }
+        stateControllerUI->setProperty("state", state);
+    }
```

Comment:
This would result in an application crash, which is undesired. Getting and setting state is only meant to update the UI and should not crash if the UI is not ready yet.